_source/_posts/2021-01-20-reactive-java-microservices.adoc

```diff
@@ -0,0 +1,545 @@
+
+In production, you might not want to manage your own identity provider instance. That's where Okta comes in. We're a developer-friendly SaaS company that provides OAuth and OIDC support as a service. You can create a developer account and use it for free. Our https://cli.okta.com[Okta CLI] makes it super easy.
+
+Install the Okta CLI using the aforementioned website and come back here when you're done. If you don't have an Okta developer account, run `okta register`.
+
+NOTE: You can also use your browser and Okta's developer console to register an app. See https://www.jhipster.tech/security/#okta[JHipster's security documentation] for those instructions.
+
+Then, from the gateway project's directory, run `okta apps create jhipster`. Accept the default redirect URIs.
+
+image::{% asset_path 'blog/reactive-java-microservices/okta-apps-create.png' %}[alt=Okta CLI in action,width=800,align=center]
+
+This process does several things:
+
+. Registers an OIDC app in Okta with JHipster's configured redirect URIs.
+. Creates `ROLE_ADMIN` and `ROLE_USER` groups and adds your user to both.
+. Creates a `groups` claim and adds it to ID tokens.
+. Creates a `.okta.env` file with the values you'll need to talk to Okta.
+
+=== Update the JHipster Registry to Distribute OIDC Configuration
+
+I mentioned earlier that Spring Cloud Config allows you to distribute Spring's configuration between apps. In this section, you'll configure JHipster's Spring Security settings to use Okta across all your services.
+
+Add the following YAML to `gateway/src/main/docker/central-server-config/localhost-config/application.yml`. You can find the values for each property in the `.okta.env` file.
+
+[source,yaml]
+----
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://<your-okta-domain>/oauth2/default
+        registration:
+          oidc:
+            client-id: <client-id>
+            client-secret: <client-secret>
+----
+
+Save your changes. These values will be distributed to the JHipster Registry, gateway, blog, and store apps. Restart the JHipster Registry by running the following commands:
+
+[source,shell]
+----
+docker-compose -f src/main/docker/jhipster-registry.yml down #jhregistrydown
+docker-compose -f src/main/docker/jhipster-registry.yml up -d #jhregistryup
+----
+
+Use kbd:[Ctrl + C] to kill all your `./gradlew` processes and start them again.
+
+Now, open a new incognito browser window, go to `http://localhost:8080`, and sign in. Rejoice that using Okta for authentication works!
+
+image::{% asset_path 'blog/reactive-java-microservices/okta-login.png' %}[alt=Okta login,width=800,align=center]
+
+image::{% asset_path 'blog/reactive-java-microservices/okta-login-success.png' %}[alt=Okta login success,width=800,align=center]
+
+If you're feeling lucky, you can set your Okta credentials as environment variables and run end-to-end tests (from the `gateway` directory).
```

Comment:
This is cool!  maybe in the future, we could support creating an Okta using the API/SDK.

Reply:
What do you mean? `okta apps create jhipster` is pretty good now!

Reply:
The E2E test part portion it needs a user/pass? I'm assuming it drives a browser.  Instead of using a static account we _could_ create one before the tests run. (not a 1 size fits all solution though, we can chat about it later)

Reply:
That would be pretty slick. The JHipster team is interested in integrating Okta into our CI processes (not in every PR, but maybe nightly). 

Have you tried the CLI in a CI process? It seems like you could set the token as a secret, which would allow talking to the Okta API.

It'd be cool to set up CI for the okta-samples repos too!

Reply:
Good call! Most of the Okta CLI ITs, mock out user input (via standard in). Not all of the options are available in a batch mode yet (I was just waiting for a use case for it 😄)